cirq-core/cirq/sim/act_on_state_vector_args.py

```diff
@@ -251,9 +254,13 @@ def prepare_into_buffer(k: int):
     if p >= 0 or weight == 0:
         # Floating point error resulted in a malformed sample.
         # Fall back to the most likely case.
-        prepare_into_buffer(fallback_weight_i)
+        prepare_into_buffer(fallback_weight_index)
         weight = fallback_weight
+        index = fallback_weight_index
 
     args.available_buffer /= np.sqrt(weight)
     args.swap_target_tensor_for(args.available_buffer)
+    if protocols.is_measurement(action):
```

Comment:
Thinking out loud. Gut feeling is there's a better way to abstract this. It seems arbitrary to have this right here for this one use case, right? Seems like there should be a way such that any op with `_measurement_key_` should be able to record that result, regardless of strategy chosen, without sprinkling this kind of logic everywhere.

One thing I was thinking is that eventually I'd like `protocols.act_on(op, args)` to return the dictionary of measurement results, rather than hiding them in the `log_of` behind the scenes. Or maybe in addition to. (Primarily because I'd like to have our iterator explicitly maintain this for each moment, rather than clearing the log after each moment, which seems like a violation of encapsulation; the log should be readonly externally IMO). Anyway, maybe this could be related to an abstraction of this generalized measurements scheme. IDK, like I said, thinking out loud.

Reply:
Slept on this. I guess what it comes down to is there's nowhere else to put it. This particular strategy is the only thing that holds this particular information, so.

The one thing I do still question is whether we want to overload "measurements" here. To me, measurement has a very specific meaning in the context of quantum computing. Overloading it like this could cause confusion. To me, this is more like "trace" information, and perhaps should go somewhere else with a different format.

Additionally, I think we need to consider the possibility of some future use case where there's some op type where we want to provide multiple pieces of trace information. Here we only provide a single key and an int array. Of course we could pack anything into the int array we wanted to, but seems like there should be a more elegant solution.